src/plugins/plugin-koa.js

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const shimmer = require('shimmer');
+var urlParse = require('url').parse;
+
+const SUPPORTED_VERSIONS = '1.x';
+
+var api;
+
+function useWrap(use) {
+  return function useTrace() {
+    if (!this._google_trace_patched) {
+      this._google_trace_patched = true;
+      this.use(middleware);
+    }
+    return use.apply(this, arguments);
+  };
+}
+
+function* middleware(next) {
+  /* jshint validthis:true */
+  const req = this.req;
+  const res = this.res;
+  const originalEnd = res.end;
+  var options = {
+    name: urlParse(req.url).pathname,
+    traceContext: this.req.headers[api.constants.TRACE_CONTEXT_HEADER_NAME],
+    skipFrames: 3
+  };
+  api.runInRootSpan(options, function(root) {
+    if (!root) {
+      return;
+    }
+
+    api.wrapEmitter(req);
+    api.wrapEmitter(res);
+
+    const url = (req.headers['X-Forwarded-Proto'] || 'http') +
+      '://' + req.headers.host + req.url;
+    
+    // we use the path part of the url as the span name and add the full
+    // url as a label
+    // req.path would be more desirable but is not set at the time our middlewear runs.
+    root.addLabel(api.labels.HTTP_METHOD_LABEL_KEY, req.method);
+    root.addLabel(api.labels.HTTP_URL_LABEL_KEY, url);
+    root.addLabel(api.labels.HTTP_SOURCE_IP, req.connection.remoteAddress);
+
+    var context = root.getTraceContext();
+    if (context) {
+      res.setHeader(api.constants.TRACE_CONTEXT_HEADER_NAME, context);
+    }
+    //else {
+    //  TODO: Determine if this is still needed
+    //  agent.logger.warn('koa: Attempted to generate trace context for nullSpan');
```

Comment:
This log is no longer needed since `runInRootSpan` returns `null` now if a span should not be created instead of a `nullSpan`. We should think about how we want to unify this method if indicating failure with the use of `nullSpan` on the existing instrumentation api. We originally used `nullSpan` so that people instrumenting their own code wouldn't need to wrap all operations on spans in a null check. @kjin just want to put this on your radar.